_posts/2018-05-23-getting-foreman-search-results-into-your-puppet-manifest.md

```diff
@@ -0,0 +1,98 @@
+
+## An use case out of standard usages.
+
+Use cases about monitoring or backups are standard. Lets try to build another example where we can use the Foreman search engine.
+
+Imagine that you are managing nodes with a Puppet Server and they are in `hostgroups` like `secure_base/role`. Where `secure_base` permits to apply all common settings to all your nodes in a secured environment. And `role` permits to apply settings dedicated to the role of a node. Now, lets say you have the constraint to not use a DNS server. And your customer would like to call hosts in a human way, so with names and not with IPs. I already met this use case. Fortunately, servers were using only one IP address.
+
+So you have to manage your `/etc/hosts` with a configuration management tool or you become crazy. 
+
+If each Puppet node execute the following code :
+
+```
+  $query = foreman({foreman_user => 'apiuser',
+                    foreman_pass => 'here_a_password',
+                    item         => 'hosts',
+                    search       => 'hostgroup_fullname ~ secure_base',
+  })
+
+  $nodes = $query['results']
+
+```
+
+In `$nodes`, you got all the needed informations about other nodes. You can check my statement by adding the following piece of code :
+
+
+```
+  $nodes.each | $_node | {
+    $_thiscert = $_node['certname']
+    $_thisip = $_node['ip']
+    notify {  "node ${_thiscert}" : message => "host ${_thisip}"}
+  }
+
+```
+
+Let's use `$nodes` in a template with a `file` resource to manage `/etc/hosts` and you got the trick.
+
+## Unknow function foreman()
+
+If you ran the code above, you probably got an error message like `Evaluation Error: Unknown function: 'foreman'`. By default, the function `foreman()` is not available.
+
+Several possibilities to get the function into your environement :
+
+ * You installed Foreman with `foreman-installer`. And so, you can simply copy `/usr/share/foreman-installer/modules/foreman/lib/puppet/parser/functions/foreman.rb` into your module managing `/etc/hosts`.
+ * Otherwise, you can [download the function foreman()](https://github.com/theforeman/puppet-foreman/blob/master/lib/puppet/parser/functions/foreman.rb) and use it into your module managing `/etc/hosts`.
+ * Probably a better way is to create a module named like `foreman-libs` containing the function `foreman()`. Add this module in your list of imported modules. 
+
+It is also possible to write your own function to query [the API](https://theforeman.org/documentation) in a way that fits your needs.
+
+## A look at foreman()'s code
+
+The code is clearly commented. Only by reading comments, it is possible to add following informations.
+
+We used option `item` with `hosts`. But, it may be `environments`, `fact_values`, `hosts`, `hostgroups`, `puppetclasses`, `smart_proxies` or `subnets`.
+
+We didn't used option `per_page`. By default, the function returns only 20 items in a row.
```

Comment:
The default `per_page` is a setting in your foreman installation, but that defaults to 20 items. It might also be good to show an example of how to override this.

Reply:
As you wish for the example.

But i don't understand your comment : 
> The default per_page is a setting in your foreman installation

Reply:
I meant to say the 20 is not hard coded but can vary per installation. But I think the by default is fine.